File: app/services/poc_video_url.py
"""Allowlisted HTTPS video résumé URLs for Jobs CRM apply.

Paste Loom / YouTube / Vimeo (embed) or Google Drive (link-out).
No video file upload. Empty is valid — PoC stays skippable (F11).
Error strings must not echo the pasted URL.
"""
from __future__ import annotations

import re
from urllib.parse import parse_qs, urlparse

POC_VIDEO_MAX_LEN = 2000

POC_VIDEO_BAD_SCHEME = (
    "Paste an HTTPS Loom, YouTube, Vimeo, or Google Drive link. "
    "Empty is fine — this does not block apply."
)
POC_VIDEO_BAD_HOST = (
    "That host is not allowed. Use Loom, YouTube, Vimeo, or Google Drive."
)
POC_VIDEO_TOO_LONG = "Video URL is too long."

_EMBED_KINDS = ("loom", "youtube", "vimeo")
_YOUTUBE_ID = re.compile(r"^[A-Za-z0-9_-]{11}$")
_LOOM_ID = re.compile(r"^[A-Za-z0-9-]{8,64}$")
_VIMEO_ID = re.compile(r"^(\d{5,12})$")


def _bare_host(host: str) -> str:
    h = (host or "").lower().rstrip(".")
    if h.startswith("www."):
        h = h[4:]
    return h


def _host_matches(host: str, suffixes: tuple[str, ...]) -> bool:
    h = _bare_host(host)
    for suffix in suffixes:
        if h == suffix or h.endswith("." + suffix):
            return True
    return False


def classify_poc_video_host(host: str) -> str | None:
    if _host_matches(host, ("youtu.be", "youtube.com")):
        return "youtube"
    if _host_matches(host, ("loom.com",)):
        return "loom"
    if _host_matches(host, ("vimeo.com",)):
        return "vimeo"
    if _host_matches(host, ("drive.google.com",)):
        return "drive"
    return None


def normalize_poc_video_url(raw: str | None) -> str | None:
    """Return a cleaned HTTPS URL, or None when empty.

    Raises ValueError on a non-empty invalid value. Messages omit the URL.
    """
    text = (raw or "").strip()
    if not text:
        return None
    if len(text) > POC_VIDEO_MAX_LEN:
        raise ValueError(POC_VIDEO_TOO_LONG)
    parsed = urlparse(text)
    if parsed.scheme.lower() != "https":
        raise ValueError(POC_VIDEO_BAD_SCHEME)
    if parsed.username or parsed.password:
        raise ValueError(POC_VIDEO_BAD_SCHEME)
    host = parsed.hostname or ""
    if not classify_poc_video_host(host):
        raise ValueError(POC_VIDEO_BAD_HOST)
    cleaned = parsed._replace(netloc=host.lower(), fragment="").geturl()
    return cleaned


def poc_video_kind(url: str | None) -> str | None:
    text = (url or "").strip()
    if not text:
        return None
    try:
        parsed = urlparse(text)
    except ValueError:
        return None
    if parsed.scheme.lower() != "https":
        return None
    return classify_poc_video_host(parsed.hostname or "")


def poc_video_embed_url(url: str | None) -> str | None:
    """Safe iframe src for Loom / YouTube / Vimeo. Drive is link-out only."""
    text = (url or "").strip()
    if not text:
        return None
    try:
        parsed = urlparse(text)
    except ValueError:
        return None
    if parsed.scheme.lower() != "https":
        return None
    kind = classify_poc_video_host(parsed.hostname or "")
    if kind not in _EMBED_KINDS:
        return None
    host = _bare_host(parsed.hostname or "")
    parts = [p for p in (parsed.path or "").split("/") if p]
    query = parse_qs(parsed.query or "")

    if kind == "youtube":
        vid = ""
        if host == "youtu.be" and parts:
            vid = parts[0]
        elif "v" in query and query["v"]:
            vid = query["v"][0]
        elif parts and parts[0] in {"embed", "shorts", "v"} and len(parts) > 1:
            vid = parts[1]
        vid = (vid or "").split("?")[0].split("&")[0]
        if not _YOUTUBE_ID.match(vid):
            return None
        return f"https://www.youtube-nocookie.com/embed/{vid}"

    if kind == "loom":
        share = ""
        if parts and parts[0] in {"share", "embed"} and len(parts) > 1:
            share = parts[1]
        elif parts:
            share = parts[-1]
        share = (share or "").split("?")[0]
        if not _LOOM_ID.match(share):
            return None
        return f"https://www.loom.com/embed/{share}"

    if kind == "vimeo":
        vid = ""
        if parts and parts[0] == "video" and len(parts) > 1:
            vid = parts[1]
        else:
            for part in reversed(parts):
                if _VIMEO_ID.match(part):
                    vid = part
                    break
        if not _VIMEO_ID.match(vid or ""):
            return None
        return f"https://player.vimeo.com/video/{vid}"

    return None

File: readyforrobots-new/client/src/lib/pocVideoUrl.ts
/**
 * Allowlisted HTTPS video résumé URLs for Jobs CRM apply.
 * Loom / YouTube / Vimeo embed; Google Drive is a link-out. Empty is valid.
 * Error strings must not echo the pasted URL.
 */

export const JOBS_POC_VIDEO_LABEL = "Async video résumé for this Job Card";
export const JOBS_POC_VIDEO_HINT =
  "Paste an unlisted Loom, YouTube, or Vimeo URL. Google Drive opens as a link, not an embed. Empty is fine — this does not block apply. Do not upload a video file here.";
export const JOBS_POC_VIDEO_WATCH = "Watch demo";
export const JOBS_POC_VIDEO_SCRIPT_HEADING =
  "Record this Job Card, not a brand reel";
export const JOBS_POC_VIDEO_BAD_SCHEME =
  "Paste an HTTPS Loom, YouTube, Vimeo, or Google Drive link. Empty is fine — this does not block apply.";
export const JOBS_POC_VIDEO_BAD_HOST =
  "That host is not allowed. Use Loom, YouTube, Vimeo, or Google Drive.";

export type PocVideoKind = "loom" | "youtube" | "vimeo" | "drive";

export type PocVideoParsed = {
  url: string;
  kind: PocVideoKind;
  embedUrl: string | null;
};

export type PocVideoScriptBeat = {
  n: number;
  title: string;
  body: string;
};

const YOUTUBE_ID = /^[A-Za-z0-9_-]{11}$/;
const LOOM_ID = /^[A-Za-z0-9-]{8,64}$/;
const VIMEO_ID = /^(\d{5,12})$/;
const MAX_LEN = 2000;

function bareHost(host: string): string {
  let h = (host || "").toLowerCase().replace(/\.$/, "");
  if (h.startsWith("www.")) h = h.slice(4);
  return h;
}

function hostMatches(host: string, suffixes: string[]): boolean {
  const h = bareHost(host);
  return suffixes.some(suffix => h === suffix || h.endsWith(`.${suffix}`));
}

export function classifyPocVideoHost(host: string): PocVideoKind | null {
  if (hostMatches(host, ["youtu.be", "youtube.com"])) return "youtube";
  if (hostMatches(host, ["loom.com"])) return "loom";
  if (hostMatches(host, ["vimeo.com"])) return "vimeo";
  if (hostMatches(host, ["drive.google.com"])) return "drive";
  return null;
}

function youtubeId(parsed: URL): string {
  const host = bareHost(parsed.hostname);
  const parts = parsed.pathname.split("/").filter(Boolean);
  if (host === "youtu.be" && parts[0]) return parts[0].split("&")[0];
  const v = parsed.searchParams.get("v");
  if (v) return v;
  if (
    (parts[0] === "embed" || parts[0] === "shorts" || parts[0] === "v") &&
    parts[1]
  ) {
    return parts[1];
  }
  return "";
}

function loomId(parsed: URL): string {
  const parts = parsed.pathname.split("/").filter(Boolean);
  if ((parts[0] === "share" || parts[0] === "embed") && parts[1])
    return parts[1];
  return parts[parts.length - 1] || "";
}

function vimeoId(parsed: URL): string {
  const parts = parsed.pathname.split("/").filter(Boolean);
  if (parts[0] === "video" && parts[1]) return parts[1];
  for (let i = parts.length - 1; i >= 0; i -= 1) {
    if (VIMEO_ID.test(parts[i])) return parts[i];
  }
  return "";
}

export function parsePocVideoUrl(
  raw: string | null | undefined
): PocVideoParsed | null {
  const text = (raw || "").trim();
  if (!text) return null;
  if (text.length > MAX_LEN) {
    throw new Error("Video URL is too long.");
  }
  let parsed: URL;
  try {
    parsed = new URL(text);
  } catch {
    throw new Error(JOBS_POC_VIDEO_BAD_SCHEME);
  }
  if (parsed.protocol !== "https:") {
    throw new Error(JOBS_POC_VIDEO_BAD_SCHEME);
  }
  if (parsed.username || parsed.password) {
    throw new Error(JOBS_POC_VIDEO_BAD_SCHEME);
  }
  const kind = classifyPocVideoHost(parsed.hostname);
  if (!kind) {
    throw new Error(JOBS_POC_VIDEO_BAD_HOST);
  }
  parsed.hash = "";
  const url = parsed.toString();
  let embedUrl: string | null = null;
  if (kind === "youtube") {
    const id = youtubeId(parsed);
    if (YOUTUBE_ID.test(id)) {
      embedUrl = `https://www.youtube-nocookie.com/embed/${id}`;
    }
  } else if (kind === "loom") {
    const id = loomId(parsed);
    if (LOOM_ID.test(id)) {
      embedUrl = `https://www.loom.com/embed/${id}`;
    }
  } else if (kind === "vimeo") {
    const id = vimeoId(parsed);
    if (VIMEO_ID.test(id)) {
      embedUrl = `https://player.vimeo.com/video/${id}`;
    }
  }
  return { url, kind, embedUrl };
}

/** Empty is not an issue. Invalid non-empty input returns a host/scheme message (no URL). */
export function pocVideoUrlIssue(
  raw: string | null | undefined
): string | null {
  const text = (raw || "").trim();
  if (!text) return null;
  try {
    parsePocVideoUrl(text);
    return null;
  } catch (err) {
    return err instanceof Error ? err.message : JOBS_POC_VIDEO_BAD_SCHEME;
  }
}

export function pocVideoScriptBeats(opts: {
  robotName: string;
  selectedModels?: string[];
  employer?: string | null;
  jobTitle?: string;
  work?: string;
  requirements?: string[];
}): PocVideoScriptBeat[] {
  const robot = (opts.robotName || "this robot").trim() || "this robot";
  const models = (opts.selectedModels || []).map(m => m.trim()).filter(Boolean);
  const skuLine = models.length
    ? models.join(", ")
    : "the catalogued SKU you will use";
  const employer = (opts.employer || "this employer").trim() || "this employer";
  const jobTitle = (opts.jobTitle || "this job").trim() || "this job";
  const work = (opts.work || "").trim();
  const reqs = (opts.requirements || []).map(r => r.trim()).filter(Boolean);
  const reqLine = reqs.length ? reqs.slice(0, 3).join("; ") : work || jobTitle;
  return [
    {
      n: 1,
      title: "Robot + SKU",
      body: `Show ${robot} and ${skuLine} so they can see which machine this is.`,
    },
    {
      n: 2,
      title: "This employer’s work",
      body: `Show the work ${employer} needs on “${jobTitle}”: ${reqLine}.`,
    },
    {
      n: 3,
      title: "60–90s demo",
      body: "Keep it to 60–90 seconds. Unlisted link. This is a résumé for this Job Card, not a brand reel.",
    },
  ];
}
